### PR TITLE
Test enkf_linalg_genX2 with catch

### LINKED ISSUE #2346: Test `enkf_linalg_genX2` using `catch2`
Introduce `catch2` by writing a test of `enkf_linalg_genX2`.

### PR DESCRIPTION
**Pull request by dafeda**
**Issue**
Resolves #2346

### CODE REVIEW

**Comment by dafeda** [comment]
Blocked until catch2 has been added via Conan by this PR: https://github.com/equinor/ert/pull/2350

File: libres/lib/analysis/enkf_linalg.cpp [hunk 1]
```diff
@@ -35,10 +35,28 @@ void enkf_linalg_genX3(matrix_type *X3, const matrix_type *W,
     matrix_free(X2);
 }
 
+/*
```
Comment (dotfloat) [inline libres/lib/analysis/enkf_linalg.cpp:38]: C++ comments tend to have the following form:
```c++
/**
 * [ TEXT ]
 */
```

We are sort of trying to use doxygen for documentation generation, although it's not actively being generated. We can still attempt to write docs in that format though.

File: libres/lib/analysis/enkf_linalg.cpp [hunk 2]
```diff
@@ -35,10 +35,28 @@ void enkf_linalg_genX3(matrix_type *X3, const matrix_type *W,
...
+    X2 = (I + \Lambda_1)^{-1/2} @ X^T_1 * S
+
+, but this function seems to implement the following:
+
+    X2 = X^T_1 * S
+    X2 = (\Lambda^{-1/2}_1 * X2^T)^T
+
+See tests for more details.
+
+TODO: The differences between equation and implementation need explaining.
```
Comment (dotfloat) [inline libres/lib/analysis/enkf_linalg.cpp:53]: Is this TODO going to be DONE in this PR? Suggest removing this line and making a separate issue on Github.

File: libres/tests/analysis/test_enkf_linalg.cpp [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+W = np.ones((nrow, ncol))
+W[0, 1] = 1.3
+
+eig = np.array([2.0, 2.0, 3.0])
+
+X2 = (W.T @ S)
+X2 = (np.sqrt(eig) * X2.T).T
+X2
+*/
+double *test_enkf_linalg_genX2() {
```
Comment (dotfloat) [inline libres/tests/analysis/test_enkf_linalg.cpp:25]: This allocates but doesn't free. Maybe you can move this into the test and use BDD?

```c++
TEST_CASE(...) {
    GIVEN("a matrix") {
        // Matrix alloc
        THEN("Use enkf_linalg_genX2") {
           enkf_linalg_genX2(...);
           REQUIRE(...);
        }
        // Matrix free
    }
}
```

File: libres/lib/analysis/enkf_linalg.cpp (hunk 2 above)
Comment (dafeda) [reply libres/lib/analysis/enkf_linalg.cpp]: Fixed, thanks.

File: libres/lib/analysis/enkf_linalg.cpp (hunk 1 above)
Comment (dafeda) [reply libres/lib/analysis/enkf_linalg.cpp]: Fixed, thanks.

File: libres/lib/analysis/enkf_linalg.cpp (hunk 2 above)
Comment (dafeda) [reply libres/lib/analysis/enkf_linalg.cpp]: Removed and created this issue: https://github.com/equinor/ert/issues/2360

Thanks

File: libres/tests/analysis/test_enkf_linalg.cpp (hunk 3 above)
Comment (dafeda) [reply libres/tests/analysis/test_enkf_linalg.cpp:25]: Is it enough to release `S` and `W`?

```C
    matrix_free(S);
    matrix_free(W);
```

Comment (dotfloat) [reply libres/tests/analysis/test_enkf_linalg.cpp:25]: You need to free `X2` also. Maybe you can change the later code to:
```cpp
    auto matrix = test_enkf_linalg_genX2();
    auto result = matrix->data;

    // ...

    matrix_free(matrix);
```

**Review by dotfloat — APPROVED** [review]
Approve4approve

File: libres/lib/analysis/enkf_linalg.cpp [hunk 4]
```diff
@@ -35,10 +35,26 @@ void enkf_linalg_genX3(matrix_type *X3, const matrix_type *W,
...
+ * The Ensemble Kalman Filter, 2nd Edition by Geir Evensen.
+ *
+ * Eq. 14.34 is as follows:
+ *
+ *     X2 = (I + \Lambda_1)^{-1/2} @ X^T_1 * S
+ *
+ * , but this function seems to implement the following:
+ *
+ *     X2 = X^T_1 * S
+ *     X2 = (\Lambda^{-1/2}_1 * X2^T)^T
```
Comment (berland) [inline libres/lib/analysis/enkf_linalg.cpp:49]: The interchange of `^{-1/2}` and `_1` compared to line 44, does that carry any meaning here?

Comment (dafeda) [reply libres/lib/analysis/enkf_linalg.cpp:49]: In line 44 we do take the square root of (I + \Lambda_1), while the code does takes the square root of just \Lambda.
Does that maybe answer your question?

**Comment by dafeda** [comment]
test jenkins please

File: libres/lib/analysis/enkf_linalg.cpp (hunk 4 above)
Comment (berland) [reply libres/lib/analysis/enkf_linalg.cpp:49]: No biggie, I think I would have written `\Lambda_1^{-1/2}` in line 49. In LaTeX it is rendered identically, but a tad less confusing for the ascii reading human.

Comment (dafeda) [reply libres/lib/analysis/enkf_linalg.cpp:49]: I cater to non-ascii readers only :)

**Pull request merged**